pkg/network/ebpf/c/protocols/tls/go-tls-location.h

```diff
@@ -101,17 +104,18 @@ static __always_inline int read_register(struct pt_regs* ctx, int64_t regnum, vo
         *(u64*)dest = tmp;
         return 0;
     #elif defined(__TARGET_ARCH_arm64)
-        // TODO Support ARM
-        /*if (regnum >= 0 && regnum < sizeof(ctx->regs)) {
-            // Verifier won't allow direct access to regs array if the index is not const
-            switch (regnum) {
-                case 0:
-                    __builtin_memcpy(dest, &ctx->regs[0], sizeof(ctx->regs[0]));
-                default:
-                    return 1;
+        volatile u64 tmp = 0;
+    #pragma unroll
+        for (int i = 0; i < NUM_REGISTERS; i++) {
+            if (i == regnum) {
+                tmp = ctx->regs[i];
             }
+        }
```

Comment:
why can we simply do `tmp = ctx->regs[regnum]`? (after a validation that `0 <= regnum && regnum < NUM_REGISTERS`)

Reply:
in this case the problem is actually not in the bound/range check, but rather that the verifier simply doesn't allow a _variable_ access of the ctx pointer (it throws the error `variable ctx access var_off=(0x0; ...) disallowed`
see https://elixir.bootlin.com/linux/v5.10/source/kernel/bpf/verifier.c#L3146
doing the unrolled loop ensures that the offsets are constant/known at load time

Reply:
I'm wondering what am I missing, why can't we do it here, but we can in `read_register_indirect`

Reply:
`read_register_indirect` does only some pointer arithmetic and returns a memory address. the actual "read" is then done via the `bpf_probe_read` helper, and so there is no direct ctx pointer derefence as in `read_register`.

Reply:
to be honest I'm not sure why have the 2 functions, since both do exactly the same thing (I was just following the implementation of x86), but I guess I can open a folllow-up PR and simplify that